Yosemite/Yosemite/Stores/ProductTagStore.swift

```diff
@@ -0,0 +1,195 @@
+import Foundation
+import Networking
+import Storage
+
+// MARK: - ProductTagStore
+//
+public final class ProductTagStore: Store {
+
+    private lazy var sharedDerivedStorage: StorageType = {
+        return storageManager.newDerivedStorage()
+    }()
+
+    /// Registers for supported Actions.
+    ///
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: ProductTagAction.self)
+    }
+
+    /// Receives and executes Actions.
+    ///
+    override public func onAction(_ action: Action) {
+        guard let action = action as? ProductTagAction else {
+            assertionFailure("ProductTagStore received an unsupported action")
+            return
+        }
+
+        switch action {
+        case let .synchronizeAllProductTags(siteID, fromPageNumber, onCompletion):
+            synchronizeAllProductTags(siteID: siteID, fromPageNumber: fromPageNumber, onCompletion: onCompletion)
+        case let .addProductTags(siteID, tags, onCompletion):
+            addProductTags(siteID: siteID, tags: tags, onCompletion: onCompletion)
+        case let .deleteProductTags(siteID, ids, onCompletion):
+            deleteProductTags(siteID: siteID, ids: ids, onCompletion: onCompletion)
+        }
+    }
+}
+
+// MARK: - Services
+//
+private extension ProductTagStore {
+
+    /// Synchronizes all product tags associated with a given Site ID, starting at a specific page number.
+    ///
+    func synchronizeAllProductTags(siteID: Int64, fromPageNumber: Int, onCompletion: @escaping (ProductTagActionError?) -> Void) {
+
+        // Start fetching the provided initial page
+        synchronizeProductTags(siteID: siteID, pageNumber: fromPageNumber, pageSize: Constants.defaultMaxPageSize) { [weak self] (result) in
+            guard let self = self  else {
+                return
+            }
+
+            switch result {
+            case .success(let tags):
+                // If tags is empty, end the recursion and call `onCompletion`
+                if tags.isEmpty {
+                    onCompletion(nil)
+                    return
+                }
+
+                // Request the next page recursively
```

Comment:
Yes it is intentional. The main reason is that while the user is typing we need to show all the existing tags with that name, and we can do that correctly only fetching all the existing tags.

Reply:
ah I see, I just checked WPiOS app and it looks like it fetches all the tags at once as well.

in this case, `pageNumber` seems like an implementation detail and we probably always want to sync from the first page? would it be more clear if `pageNumber` is hidden from outside of this function, removing it from `ProductTagAction` and adding a default value to the first page number in this function? otherwise, it feels like this action is meant to work with pagination :)

Reply:
You are right, and your solution seems a really good idea! Done here 👍 6df8efcf98ea8ed45690544cde2bceaf26072939